src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/Commands/Upgrade/ConsoleUpgrade.cs

```diff
@@ -76,9 +86,32 @@ public async Task RunAsync(CancellationToken token)
             {
                 // Do not pass the same token as it may have been canceled and we still need to persist this.
                 await _stateManager.SaveStateAsync(context, default);
+                await WriteUpgradeReport(context, token);
             }
         }
 
+        private async Task<bool> WriteUpgradeReport(IUpgradeContext context, CancellationToken token)
+        {
+            var outputDirectory = Path.GetDirectoryName(context.InputPath);
+            if (outputDirectory is null || !context.Results.Any())
+            {
+                return false;
+            }
+
+            var outputFileName = $"UpgradeReport.{_options.Value.Format}";
+            var outputFilePath = Path.Combine(outputDirectory, outputFileName);
+            using var stream = File.Create(Path.Combine(outputDirectory, outputFileName));
+            if (_writerProvider.TryGetWriter(_options.Value.Format, out var writer))
```

Comment:
can't we verify this earlier? it would be a shame to start an upgrade, get through it, but not get a report because the format was not available

Reply:
If we aren't already, we should validate the user's choice for `--format` is one that's been registered w/ the DI container and error out if one doesn't exist. That would get at Taylor's ask here.
I _think_ we're doing this, though.